python/pyspark/sql/dataframe.py

```diff
@@ -257,53 +264,85 @@ def limit(self, num):
     @ignore_unicode_prefix
     @since(1.3)
     def take(self, num):
-        """Returns the first ``num`` rows as a :class:`list` of :class:`Row`.
+        """Returns the first ``num`` records as a :class:`list`.
 
         >>> df.take(2)
         [Row(age=2, name=u'Alice'), Row(age=5, name=u'Bob')]
         """
-        with SCCallSiteSync(self._sc) as css:
-            port = self._sc._jvm.org.apache.spark.sql.execution.python.EvaluatePython.takeAndServe(
-                self._jdf, num)
-        return list(_load_from_socket(port, BatchedSerializer(PickleSerializer())))
+        return self.limit(num).collect()
 
     @ignore_unicode_prefix
-    @since(1.3)
-    def map(self, f):
-        """ Returns a new :class:`RDD` by applying a the ``f`` function to each :class:`Row`.
+    @since(2.0)
+    def applySchema(self, schema=None):
+        """Returns a new :class:`DataFrame` by appling the given schema, or infer the schema
+        by all of the records if no schema is given.
 
-        This is a shorthand for ``df.rdd.map()``.
+        It is only allowed to apply schema for DataFrame which is returned by typed operations,
+        e.g. map, flatMap, etc. And the record type of the schema-applied DataFrame will be row.
 
-        >>> df.map(lambda p: p.name).collect()
+        >>> ds = df.map(lambda row: row.name)
+        >>> ds.collect()
         [u'Alice', u'Bob']
+        >>> ds.schema
+        StructType(List(StructField(value,BinaryType,false)))
+        >>> ds2 = ds.applySchema(StringType())
+        >>> ds2.collect()
+        [Row(value=u'Alice'), Row(value=u'Bob')]
+        >>> ds2.schema
+        StructType(List(StructField(value,StringType,true)))
+        >>> ds3 = ds.applySchema()
+        >>> ds3.collect()
+        [Row(value=u'Alice'), Row(value=u'Bob')]
+        >>> ds3.schema
+        StructType(List(StructField(value,StringType,true)))
+        """
+        msg = "Cannot apply schema to a DataFrame which is not returned by typed operations"
+        raise RuntimeError(msg)
+
+    @ignore_unicode_prefix
+    @since(2.0)
+    def mapPartitions(self, func):
+        """Returns a new :class:`DataFrame` by applying the ``f`` function to each partition.
+
+        The schema of returned :class:`DataFrame` is a single binary field struct type, please
+        call `applySchema` to set the corrected schema before apply structured operations, e.g.
+        select, sort, groupBy, etc.
+
+        >>> def f(iterator):
+        ...     return map(lambda i: 1, iterator)
+        >>> df.mapPartitions(f).collect()
+        [1, 1]
         """
-        return self.rdd.map(f)
+        return PipelinedDataFrame(self, func)
 
     @ignore_unicode_prefix
-    @since(1.3)
-    def flatMap(self, f):
-        """ Returns a new :class:`RDD` by first applying the ``f`` function to each :class:`Row`,
-        and then flattening the results.
+    @since(2.0)
```

Comment:
This API is introduced in 1.3, but changed in 2.0. We should still keep the  `since(1.3)` here, add `::versionchanged 2.0 xxx` in the doc string